python_modules/dagster/dagster/core/host_representation/external_data.py

```diff
@@ -458,45 +460,57 @@ def __new__(cls, error):
 
 @whitelist_for_serdes
 class ExternalAssetDependency(
-    namedtuple("_ExternalAssetDependency", "upstream_asset_key input_name")
+    namedtuple("_ExternalAssetDependency", "upstream_asset_key input_name output_name")
 ):
     """A definition of a directed edge in the logical asset graph.
 
     An upstream asset that's depended on, and the corresponding input name in the downstream asset
     that depends on it.
     """
 
-    def __new__(cls, upstream_asset_key: AssetKey, input_name: str):
+    def __new__(cls, upstream_asset_key: AssetKey, input_name: str = None, output_name: str = None):
+        check.invariant(
+            (input_name is None) ^ (output_name is None),
+            "Exactly one of `input_name` and `output_name` should be supplied",
+        )
         return super(ExternalAssetDependency, cls).__new__(
             cls,
             upstream_asset_key=upstream_asset_key,
             input_name=input_name,
+            output_name=output_name,
         )
 
 
 @whitelist_for_serdes
 class ExternalAssetDependedBy(
-    namedtuple("_ExternalAssetDependedBy", "downstream_asset_key input_name")
+    namedtuple("_ExternalAssetDependedBy", "downstream_asset_key input_name output_name")
 ):
     """A definition of a directed edge in the logical asset graph.
 
     An downstream asset that's depended by, and the corresponding input name in the upstream
     asset that it depends on.
     """
 
-    def __new__(cls, downstream_asset_key: AssetKey, input_name: str):
+    def __new__(
+        cls, downstream_asset_key: AssetKey, input_name: str = None, output_name: str = None
+    ):
+        check.invariant(
+            (input_name is None) ^ (output_name is None),
+            "Exactly one of `input_name` and `output_name` should be supplied",
+        )
         return super(ExternalAssetDependedBy, cls).__new__(
             cls,
             downstream_asset_key=downstream_asset_key,
             input_name=input_name,
+            output_name=output_name,
         )
 
 
 @whitelist_for_serdes
 class ExternalAssetNode(
     namedtuple(
         "_ExternalAssetNode",
-        "asset_key dependencies depended_by op_name op_description job_names partitions_def_data",
+        "asset_key dependencies depended_by op_name op_description job_names partitions_def_data output_name output_description",
```

Comment:
one thing i don't quite grok about these `ExternalAsset...` objects is why this information gets serialized. Looking at the contents I would expect them to be derivable from information already in the `ExternalRepositoryData` and these classes would be objects that `ExternalRepository` would build internally after loading `ExternalRepositoryData`

Reply:
Ahhh I think I see - the information currently on `OutputDefinition` never made its way to `OutputDefSnap` as I assumed and we encoded it here instead.